### PR TITLE
[Fix](#649) changes to support accurate ns timepoints

### PR DESCRIPTION
**Pull request by auphofBSF**
as documented in (#649). 
use floor divided on integer division and use pandas.Timestamp for ns timestamp resolution

Extensively tested but not fully tested, unit test updates to follow
I believe address and fixes ( #527,#489, #346, #344, #340)

### CODE REVIEW

**Comment by aviau** [comment]
I'd be happy to merge if it includes unit tests.

**Comment by auphofBSF** [comment]
Have added fixes to enable passing existing python unit tests.  
#### However some failures persist

1. Unit Test failures are with pypy  because of the now mandatory dependency on pandas to do ns precision timestamps. I have no experience with pypy can we allow dependency on pandas for pypy ?
2. The other unit test failure is flake8,  I cannot figure this one out HELP please someone


I am testing further unit test updates and additions for confirmation of ns precision conformance.

**Comment by tpietruszka** [comment]
While I believe this PR improves the situation, I still see a problem.

**When trying to write points with microsecond precision, the timestamps are sometimes erroneous (1 us too early)**

Why? **The `precision_factor` values (other than 1), in `_dataframe_client.py`, are floats, and so the result of floor division is also a float.** For microsecond precision this poses a problem - most timestamps only have approximate float representations. If the approximation is smaller than the actual value, the next call to `astype(np.int64)` takes a floor value of something smaller than the value we want.

**Solution: 
`precision_factor` values should be ints. BTW, maybe they could be defined just once.**



Example:
```
import pandas as pd
import influxdb

test_db = 'test_python_client_timestamp_precision'
col_name = 'temperature'
measurement = 'test'

df = pd.Series({pd.Timestamp('2019-04-16 16:24:56.915728Z'): 1}, name = col_name).to_frame()

client = influxdb.DataFrameClient(database=test_db)
client.create_database(test_db)
client.write_points(df, measurement, time_precision='u')

df2 = client.query(f'select * from {measurement}')[measurement]
client.drop_database(test_db)
client.close()
pd.testing.assert_index_equal(df.index, df2.index)
```
**(the assert fails both in current master AND in a version including this PR)**

**Comment by auphofBSF** [comment]
The issue identified by @tpietruszka  https://github.com/influxdata/influxdb-python/pull/650#issuecomment-485915261 is valid 

It happens on certain values, I am not well experienced on numpy arithmetic particular corner cases of conversion to challenge what is correct or incorrect. But @tpietruszka code and the simple experiment of test values and different arithmetic strategies for conversion recorded below demonstrate the issue.  I have changed the test values in the Unit tests to 

>  '2013-01-01 23:10:55.123456987+00:00' 

This change to unit tests is in the following branch https://github.com/auphofBSF/influxdb-python/tree/bug/microsecond_issue_tpietruszka.
 **Certainly all tests fail.**  


The  @tpietruszka https://github.com/influxdata/influxdb-python/pull/650#issuecomment-485915261 result should be identical (There are no NanoSeconds)  but where there are nanoseconds what to do ?

### A QUESTION and DECISION require about Rounding or Truncating
In the following experiment where there are NanoSecond components of the time should the value be truncated to the MicroSecond or Rounded Up to the Microsecond and for that matter every conversion precision.  

I am happy to assist with further alterations, thanks to @tpietruszka for identifying and solving to this stage

image


```python
ts = pd.Timestamp('2013-01-01 23:10:55.123456987+00:00')
ts
ts_ns = np.int64(ts.value)

ts_ns
expected_ts_us = 1357081855123456

#as before Pull Request #650
result1 = np.int64(ts_ns / precision_factor)
(expected_ts_us == result1, result1)

# as in Pull Request #650
result2 = np.int64(ts_ns // precision_factor)
(expected_ts_us == result2, result2)


# Tests for Fix identified by @tpietruszka  DIVIDE
result3=np.int64(ts_ns / np.int64(precision_factor))
(expected_ts_us == result3, result3)

# Tests for Fix identified by @tpietruszka  FLOOR DIVIDE
result4=np.int64(ts_ns // np.int64(precision_factor))
(expected_ts_us == result4, result4)

```

File: influxdb/line_protocol.py [hunk 1]
```diff
@@ -9,37 +9,45 @@
 from datetime import datetime
 from numbers import Integral
 
-from pytz import UTC
-from dateutil.parser import parse
 from six import iteritems, binary_type, text_type, integer_types, PY2
 
-EPOCH = UTC.localize(datetime.utcfromtimestamp(0))
+import pandas as pd  # Provide for ns timestamps
```
Comment (lipi) [inline influxdb/line_protocol.py]: Please add pandas to requirements.txt (and remove pytz and python-dateutil), build is currently failing because of that, see pypy build logs: https://travis-ci.org/influxdata/influxdb-python/jobs/516866320

Great fix btw, looking forward to being merged.

**Comment by auphofBSF** [comment]
The commits now include fixes for @tpietruszka  identified issues with precision conversion. Test cases have been updated and pass. 

The only failures still exist with pypy.   Following @lipi recommendation in https://github.com/influxdata/influxdb-python/pull/650#discussion_r292698794, enabled pypy3 to pass but then failed on all standard python implementations  clash of  tox.ini deps: and requirements.txt  duplicate requirements. 
Additionally pypy2.7 fails with pandas having a dependency on numpy 1.17 which only supports python >3.5.
I have a seperate local branch https://github.com/auphofBSF/influxdb-python/tree/test/pypy27pandas where I will explore fixing this

@lipi 
1) do you have ANY further suggestions regarding PYPY2.7 and pandas
2) how best to add pandas to requirements and have CI tests still use there depend in tox.ini

**Comment by clslgrnc** [comment]
#407 seemed much more lightweight to solve the problem. I am not sure why panda is needed just for time conversion, and I do not really understand why all the powers of 10 are defined as floats with the notation 1eX (which result in a float in python) then converted to numpy 64 bits ints. Why not just use 10 ** X which is already a python integer with arbitrary precision?

**Comment by sebito91** [comment]
If this still fails `pypy` inclusion I will re-address #407.

**Comment by sebito91** [comment]
Duplicate of #407 without `pandas` requirement
